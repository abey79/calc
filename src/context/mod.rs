pub mod ast;
pub mod checked_ast;
pub mod source;
pub mod token_stream;
