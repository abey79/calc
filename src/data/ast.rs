//! This module contains the AST data structures.
//!
//! Throughout the code, the "Node" terminology is used to refer to elements of the AST. AST nodes
//! are assigned a unique `NodeId` when created. Factory functions are also provided to make it easy
//! to build AST nodes.

use crate::data::identified::{new_id, Identified};
use crate::data::token::TokenKind;
use std::fmt;

// =================================================================================================
// DATA STRUCTURES

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, PartialOrd, Ord)]
pub struct NodeId(usize);

impl NodeId {
    pub fn new() -> Self {
        Self(new_id())
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum UnaryOpKind {
    Pos,
    Neg,
}

pub type UnaryOp = Identified<UnaryOpKind, NodeId>;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum BinOpKind {
    Add,
    Sub,
    Mul,
    Div,
}

pub type BinOp = Identified<BinOpKind, NodeId>;

pub type VarName = Identified<String, NodeId>;

#[derive(Debug, Clone, PartialEq)]
pub enum ExprKind {
    Variable(VarName),
    BinOp {
        op: BinOp,
        left: Box<Expr>,
        right: Box<Expr>,
    },
    UnaryOp {
        op: UnaryOp,
        operand: Box<Expr>,
    },
    Integer(i32),
    Float(f64),
}

pub type Expr = Identified<ExprKind, NodeId>;

#[derive(Debug, Clone, PartialEq)]
pub enum StmtKind {
    Assign { name: VarName, value: Expr },
    Print { expr: Expr },
    Expr { expr: Expr },
}

pub type Stmt = Identified<StmtKind, NodeId>;

#[derive(Debug, Clone, PartialEq)]
pub struct BlockKind {
    pub stmts: Vec<Stmt>,
}

pub type Block = Identified<BlockKind, NodeId>;

// =================================================================================================
// FACTORIES

impl From<&TokenKind> for UnaryOpKind {
    fn from(value: &TokenKind) -> Self {
        match value {
            TokenKind::Plus => Self::Pos,
            TokenKind::Minus => Self::Neg,
            _ => panic!("Invalid token kind: {:?}", value),
        }
    }
}

impl UnaryOp {
    pub fn new(kind: impl Into<UnaryOpKind>) -> Self {
        Self {
            kind: kind.into(),
            id: NodeId::new(),
        }
    }
}

impl From<&TokenKind> for BinOpKind {
    fn from(value: &TokenKind) -> Self {
        match value {
            TokenKind::Plus => Self::Add,
            TokenKind::Minus => Self::Sub,
            TokenKind::Star => Self::Mul,
            TokenKind::Slash => Self::Div,
            _ => panic!("Invalid token kind: {:?}", value),
        }
    }
}

impl BinOp {
    pub fn new(kind: impl Into<BinOpKind>) -> Self {
        Self {
            kind: kind.into(),
            id: NodeId::new(),
        }
    }
}

impl VarName {
    pub fn new(name: impl Into<String>) -> Self {
        Self {
            kind: name.into(),
            id: NodeId::new(),
        }
    }
}

impl Expr {
    pub fn variable(name: impl Into<VarName>) -> Self {
        Self {
            kind: ExprKind::Variable(name.into()),
            id: NodeId::new(),
        }
    }

    pub fn bin_op(op: BinOp, left: Expr, right: Expr) -> Self {
        Self {
            kind: ExprKind::BinOp {
                op,
                left: Box::new(left),
                right: Box::new(right),
            },
            id: NodeId::new(),
        }
    }

    pub fn unary_op(op: UnaryOp, operand: Expr) -> Self {
        Self {
            kind: ExprKind::UnaryOp {
                op,
                operand: Box::new(operand),
            },
            id: NodeId::new(),
        }
    }

    pub fn integer(value: i32) -> Self {
        Self {
            kind: ExprKind::Integer(value),
            id: NodeId::new(),
        }
    }

    pub fn float(value: f64) -> Self {
        Self {
            kind: ExprKind::Float(value),
            id: NodeId::new(),
        }
    }
}

impl Stmt {
    pub fn assign(name: impl Into<VarName>, value: Expr) -> Self {
        Self {
            kind: StmtKind::Assign {
                name: name.into(),
                value,
            },
            id: NodeId::new(),
        }
    }

    pub fn print(expr: Expr) -> Self {
        Self {
            kind: StmtKind::Print { expr },
            id: NodeId::new(),
        }
    }

    pub fn expr(expr: Expr) -> Self {
        Self {
            kind: StmtKind::Expr { expr },
            id: NodeId::new(),
        }
    }
}

impl Block {
    pub fn new(stmts: Vec<Stmt>) -> Self {
        Self {
            kind: BlockKind { stmts },
            id: NodeId::new(),
        }
    }
}

// =================================================================================================
// UTILITY TRAITS

impl fmt::Display for NodeId {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        self.0.fmt(f)
    }
}

impl fmt::Display for BinOpKind {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::Add => write!(f, "+"),
            Self::Sub => write!(f, "-"),
            Self::Mul => write!(f, "*"),
            Self::Div => write!(f, "/"),
        }
    }
}

impl fmt::Display for UnaryOpKind {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::Neg => write!(f, "-"),
            Self::Pos => write!(f, "+"),
        }
    }
}

impl From<&str> for VarName {
    fn from(s: &str) -> Self {
        Self {
            kind: s.to_string(),
            id: NodeId::new(),
        }
    }
}
