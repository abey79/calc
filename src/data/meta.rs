//! Generic type for things that have an ID, such as AST nodes.
//!
//! This type as both a generic type for the "kind" (the content of the structure) and the ID. This
//! ensure that NodeIds are not used to look up, e.g., tokens.

use std::fmt;
use std::fmt::{Debug, Display};

#[derive(Clone, PartialEq, Default)]
pub struct Meta<K, I> {
    pub kind: K,
    pub meta: I,
}

impl<K: Display, I> Display for Meta<K, I> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        self.kind.fmt(f)
    }
}

impl<K: Debug, I: Display> Debug for Meta<K, I> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "[{}] ", self.meta)?;
        self.kind.fmt(f)
    }
}

impl<K, I> AsRef<K> for Meta<K, I> {
    fn as_ref(&self) -> &K {
        &self.kind
    }
}
