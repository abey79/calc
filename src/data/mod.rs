pub mod ast;
pub mod identified;
pub mod span;
pub mod token;
