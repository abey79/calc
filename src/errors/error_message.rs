//! Error message
//!
//! This is the part of errors that contain the textual representation of the error location,
//! including a source extract and an underline.
//!
//! Built by [`Source::error_message`].

use crate::context::source::Source;
use crate::data::meta::Meta;
use crate::data::span::Span;
use std::fmt;

/// The context of an error
///
/// This is a ready-to-display string with source extract and underline of the error.
#[derive(Debug, Clone, Default, PartialEq)]
pub struct ErrorSpan(String);

impl fmt::Display for ErrorSpan {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        writeln!(f, "\n\n{}", self.0)
    }
}

impl From<&str> for ErrorSpan {
    fn from(s: &str) -> Self {
        Self(s.to_string())
    }
}

impl From<String> for ErrorSpan {
    fn from(s: String) -> Self {
        Self(s)
    }
}

/// A trait for object that have a span
///
/// A `Spanned` object somehow have a corresponding [`Span`] which can be used to construct an
/// [`ErrorSpan`], which in turn is used to display nice error messages.
pub trait Spanned {
    fn span(&self) -> Span;

    fn to_error(&self, source: &Source) -> ErrorSpan {
        if source.source().is_empty() {
            return "".into();
        }

        let span = self.span();

        let extract: String = source
            .source()
            .split('\n')
            .skip(span.start.line - 1)
            .take(span.end.line - span.start.line + 1)
            .enumerate()
            .map(|(i, line)| {
                let cur_line = i + span.start.line;
                let start = if cur_line == span.start.line {
                    span.start.col
                } else {
                    1
                };
                let end = if cur_line == span.end.line {
                    span.end.col
                } else {
                    line.len()
                };

                let underline = " ".repeat(start - 1) + &"^".repeat(end - start + 1);
                format!("{:>4} | {}\n     | {}\n", cur_line, line, underline)
            })
            .collect();

        extract.into()
    }
}

impl<K, M: Spanned> Spanned for Meta<K, M> {
    fn span(&self) -> Span {
        self.meta.span()
    }
}
