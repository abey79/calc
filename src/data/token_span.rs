use crate::data::span::Span;
use crate::data::token::Token;
use crate::errors::error_message::Spanned;
use std::fmt;
use std::rc::Rc;

#[derive(Debug, Clone)]
pub struct TokSpan {
    pub start: Rc<Token>,
    pub end: Rc<Token>,
}

impl TokSpan {
    pub fn new(start: Rc<Token>, end: Rc<Token>) -> Self {
        Self { start, end }
    }

    pub fn span(&self) -> Span {
        Span::new(self.start.span().start, self.end.span().end)
    }
}

impl fmt::Display for TokSpan {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        self.span().fmt(f)
    }
}

impl Spanned for TokSpan {
    fn span(&self) -> Span {
        self.span()
    }
}
