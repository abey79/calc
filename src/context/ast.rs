use crate::data::ast::Stmt;
use crate::data::meta::Meta;
use crate::data::span::Span;
use crate::data::token_span::TokSpan;
use crate::pipeline;
use std::fmt;
use std::fmt::{Debug, Display, Write};

#[derive(Debug, Default)]
pub struct Ast<M: Debug + Display> {
    stmts: Vec<Stmt<M>>,
}

impl<M: Debug + Display> Ast<M> {
    pub fn new() -> Self {
        Self { stmts: Vec::new() }
    }

    pub fn stmts(&self) -> &[Stmt<M>] {
        &self.stmts
    }

    pub fn stmts_mut(&mut self) -> &mut Vec<Stmt<M>> {
        &mut self.stmts
    }

    pub fn push_stmt(&mut self, stmt: Stmt<M>) {
        self.stmts.push(stmt);
    }

    pub fn format<W: fmt::Write>(&self, w: &mut W) -> Result<String, fmt::Error> {
        pipeline::formatter::format(self, w)
    }
}

impl Ast<TokSpan> {
    pub fn dump<W: Write>(&self, w: &mut W) -> std::fmt::Result {
        for stmt in self.stmts() {
            writeln!(w, "\n{:#?}", stmt)?;
        }

        Ok(())
    }
}

// Provide convenient methods for AST nodes parametrized by a token span.
impl<K> Meta<K, TokSpan> {
    pub fn tok_span(&self) -> TokSpan {
        self.meta.clone()
    }

    pub fn span(&self) -> Span {
        self.meta.span()
    }
}
