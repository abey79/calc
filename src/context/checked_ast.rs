//! State after the type checker.
//!
//! In this state, all the AST nodes are decorated with type info.

use crate::data::ast::{Expr, Stmt};
use crate::data::token_span::TokSpan;
use std::fmt;
use std::fmt::Write;

pub type CheckedStmt = Stmt<TypeInfo>;
pub type CheckedExpr = Expr<TypeInfo>;

#[derive(Debug, Default)]
pub struct CheckedAst {
    stmts: Vec<CheckedStmt>,
}

impl CheckedAst {
    pub fn new() -> Self {
        Self { stmts: Vec::new() }
    }

    pub fn stmts(&self) -> &[CheckedStmt] {
        &self.stmts
    }

    pub fn stmts_mut(&mut self) -> &mut Vec<CheckedStmt> {
        &mut self.stmts
    }

    pub fn push_stmt(&mut self, stmt: CheckedStmt) {
        self.stmts.push(stmt);
    }

    pub fn dump<W: Write>(&self, w: &mut W) -> std::fmt::Result {
        for stmt in &self.stmts {
            writeln!(w, "\n{:#?}", stmt)?;
        }

        Ok(())
    }
}

// Note:
// This is a massive simplification of what a real type structure would be. In a realistic case,
// `CheckedAst` would contain a vector of types (including user defined ones). Then, `CheckedInfo`
// would contain a ref-counted pointer to one of the types.

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum Type {
    Stmt, // stmt only
    Integer,
    Float,
}

impl fmt::Display for Type {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        use Type::*;
        match self {
            Stmt => write!(f, "stmt"),
            Integer => write!(f, "int"),
            Float => write!(f, "float"),
        }
    }
}

/// AST meta-data after type checking.
#[derive(Debug)]
pub struct TypeInfo {
    /// type of the node
    pub type_: Type,

    /// token span of the node (typically copied over from the parsed state)
    pub tok_span: TokSpan,
}

impl TypeInfo {
    pub fn new(type_: Type, tok_span: TokSpan) -> Self {
        Self { type_, tok_span }
    }
}

impl fmt::Display for TypeInfo {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.type_)
    }
}
