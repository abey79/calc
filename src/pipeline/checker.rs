use crate::context::checked_ast::{CheckedAst, CheckedExpr, CheckedStmt, Type, TypeInfo};
use crate::data::ast::{BinOp, Expr, ExprKind, Stmt, StmtKind, UnaryOp, VarName};
use crate::data::meta::Meta;
use crate::data::token_span::TokSpan;
use crate::errors::error_message::Spanned;
use crate::errors::{CheckerError, SyntaxError, TypeError};
use crate::states::{CheckedState, ParsedState};
use std::collections::HashMap;

type Result<T> = std::result::Result<T, CheckerError>;

pub(crate) fn check(input: ParsedState) -> Result<CheckedState> {
    let mut checker = Checker::new(&input);
    let checked_ast = checker.run()?;
    Ok(CheckedState {
        source: input.source,
        token_stream: input.token_stream,
        raw_ast: input.raw_ast,
        ast: checked_ast,
    })
}

struct Checker<'a> {
    input: &'a ParsedState,

    // state
    vars: HashMap<String, Type>,
}

impl<'a> Checker<'a> {
    fn new(input: &'a ParsedState) -> Self {
        Self {
            input,
            vars: HashMap::new(),
        }
    }

    fn run(&mut self) -> Result<CheckedAst> {
        let mut checked_ast = CheckedAst::new();
        for stmt in self.input.raw_ast.stmts() {
            checked_ast.push_stmt(self.check_stmt(stmt)?);
        }

        Ok(checked_ast)
    }

    fn check_stmt(&mut self, stmt: &Stmt<TokSpan>) -> Result<CheckedStmt> {
        match &stmt.kind {
            StmtKind::Assign { name, value } => {
                let checked_value = self.check_expr(value)?;
                let type_ = checked_value.meta.type_;
                self.vars.insert(name.kind.clone(), type_);
                Ok(Stmt::assign(
                    VarName::new(&name.kind, TypeInfo::new(type_, name.tok_span())),
                    checked_value,
                    TypeInfo::new(Type::Stmt, stmt.tok_span()),
                ))
            }
            StmtKind::Print { expr } => {
                let checked_expr = self.check_expr(expr)?;
                Ok(Stmt::print(
                    checked_expr,
                    TypeInfo::new(Type::Stmt, stmt.tok_span()),
                ))
            }
            StmtKind::Expr { expr } => {
                let checked_expr = self.check_expr(expr)?;
                Ok(Stmt::expr(
                    checked_expr,
                    TypeInfo::new(Type::Stmt, stmt.tok_span()),
                ))
            }
        }
    }

    fn check_expr(&mut self, expr: &Expr<TokSpan>) -> Result<CheckedExpr> {
        match &expr.kind {
            ExprKind::Variable(name) => {
                if let Some(type_) = self.vars.get(&name.kind) {
                    Ok(Expr::variable(
                        VarName::new(&name.kind, TypeInfo::new(*type_, name.tok_span())),
                        TypeInfo::new(*type_, expr.tok_span()),
                    ))
                } else {
                    Err(self.syntax_err(SyntaxError::UnknownVariable(name.kind.clone()), expr))
                }
            }
            ExprKind::BinOp { op, left, right } => {
                let checked_left = self.check_expr(left)?;
                let checked_right = self.check_expr(right)?;
                let left_type = checked_left.meta.type_;
                let right_type = checked_right.meta.type_;

                if left_type != right_type {
                    return Err(self.type_err(
                        TypeError::MismatchedTypesForBinaryOp(left_type, right_type),
                        op,
                    ));
                }

                Ok(Expr::bin_op(
                    BinOp::new(op.kind, TypeInfo::new(left_type, op.tok_span())),
                    checked_left,
                    checked_right,
                    TypeInfo::new(left_type, expr.tok_span()),
                ))
            }
            ExprKind::UnaryOp { op, operand } => {
                let checked_expr = self.check_expr(operand)?;
                let type_ = checked_expr.meta.type_;

                if !matches!(type_, Type::Integer | Type::Float) {
                    return Err(self.type_err(TypeError::InvalidTypeForUnaryOp(type_), expr));
                }

                Ok(Expr::unary_op(
                    UnaryOp::new(op.kind, TypeInfo::new(type_, op.tok_span())),
                    checked_expr,
                    TypeInfo::new(type_, expr.tok_span()),
                ))
            }
            ExprKind::Tuple(..) => todo!(),
            ExprKind::Integer(i) => Ok(Expr::integer(
                *i,
                TypeInfo::new(Type::Integer, expr.tok_span()),
            )),
            ExprKind::Float(fl) => Ok(Expr::float(
                *fl,
                TypeInfo::new(Type::Float, expr.tok_span()),
            )),
        }
    }

    fn type_err<K>(&self, err: TypeError, node: &Meta<K, TokSpan>) -> CheckerError {
        CheckerError::TypeError(err, node.to_error(&self.input.source))
    }

    fn syntax_err<K>(&self, err: SyntaxError, node: &Meta<K, TokSpan>) -> CheckerError {
        CheckerError::SyntaxError(err, node.to_error(&self.input.source))
    }
}

#[cfg(test)]
mod tests {
    use crate::pipeline::checker::check;
    use crate::pipeline::parser::parse;
    use crate::pipeline::tokenizer::tokenize;
    use crate::states::InputState;

    #[test]
    fn test_checker() {
        let input = InputState::from("a = (1.3 + 3.2) * 45.1; b = a * 3.2; print 1 + 2 * 3;");
        let tokenized = tokenize(input).unwrap();
        let parsed = parse(tokenized).unwrap();
        let checked = check(parsed).unwrap();

        insta::assert_debug_snapshot!(checked.ast);
    }
}
