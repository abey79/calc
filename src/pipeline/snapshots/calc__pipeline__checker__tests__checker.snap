---
source: src/pipeline/checker.rs
expression: checked.ast
---
Ast {
    stmts: [
        [stmt] Assign {
            name: [float] "a",
            value: [float] BinOp {
                op: [float] Mul,
                left: [float] BinOp {
                    op: [float] Add,
                    left: [float] Float(
                        1.3,
                    ),
                    right: [float] Float(
                        3.2,
                    ),
                },
                right: [float] Float(
                    45.1,
                ),
            },
        },
        [stmt] Assign {
            name: [float] "b",
            value: [float] BinOp {
                op: [float] Mul,
                left: [float] Variable(
                    [float] "a",
                ),
                right: [float] Float(
                    3.2,
                ),
            },
        },
        [stmt] Print {
            expr: [int] BinOp {
                op: [int] Add,
                left: [int] Integer(
                    1,
                ),
                right: [int] BinOp {
                    op: [int] Mul,
                    left: [int] Integer(
                        2,
                    ),
                    right: [int] Integer(
                        3,
                    ),
                },
            },
        },
    ],
}
